outlines/text/generate/integer.py

```diff
@@ -0,0 +1,72 @@
+from typing import Optional, Tuple
+
+import interegular
+import numpy as np
+import torch
+
+from outlines.text.generate.continuation import Continuation
+from outlines.text.parsing import find_partial_matches, map_partial_states_to_vocab
+
+
+class Integer(Continuation):
+    """Represents a integer generation model.
+
+    `Integer` instances are constrained generation models that only
+    generate integer values and EOS tokens. Leading zeros are fobidden.
+
+    >>> import outlines.text as text
+    >>> sequence = text.generate.integer(model)("Return an integer between 0 and 10")
+
+    """
+
+    def __init__(self, model, max_tokens: Optional[int]):
+        super().__init__(model, max_tokens)
+
+        self.vocabulary = model.tokenizer.vocabulary.keys()
+
+        int_regex_string = rf"(0|[1-9][0-9]+)({model.tokenizer.eos_token})?"
+        int_regex_pattern = interegular.parse_pattern(int_regex_string)
+        self.int_regex_fsm = int_regex_pattern.simplify().to_fsm()
+
+        def partial_match_filter(string, end_idx, state_seq):
+            if end_idx is not None and end_idx < len(string) - 1:
+                return False
+            return True
+
+        pstate_to_vocab = map_partial_states_to_vocab(
+            self.vocabulary, {"INT": self.int_regex_fsm}, True, partial_match_filter
+        )
+        self.pstate_to_vocab = {k: list(v) for k, v in pstate_to_vocab.items()}
+
+    def create_proposal(
+        self, generated_token_ids: torch.LongTensor, logits: torch.DoubleTensor
+    ) -> torch.DoubleTensor:
+        """Modify the next-token logits so that only integers can be generated.
+
+        Parameters
+        ----------
+        generated_token_ids
+            The token ids generated so far.
+        logits
+            The next-token logits.
+
+        """
+        if generated_token_ids.shape[-1] > 0:
+            sampled_sequence = self.model.tokenizer.decode(generated_token_ids)
+            partial_matches = find_partial_matches(self.int_regex_fsm, sampled_sequence)
+            pmatch = max(
+                partial_matches, key=lambda x: x[0] if x[0] is not None else -1
+            )
+            self.pstate: Tuple[str, int] = (self.pstate[0], pmatch[1][-1])
+        else:
+            self.pstate = ("INT", self.int_regex_fsm.initial)
+
+        next_support = self.pstate_to_vocab[self.pstate]
+        mask = -float("Inf") * np.ones(len(self.vocabulary), dtype=np.int64)
```

Comment:
What about `[torch|np].full_like(..., -math.inf)`?  (N.B. I'm not sure about the performance of either&mdash;`full_like` just comes to mind.)